tsl/src/continuous_aggs/invalidation_plugin.c

```diff
@@ -0,0 +1,552 @@
+		if (!OidIsValid(relinfo->hypertable_relid))
+			ereport(ERROR,
+					errcode(ERRCODE_INVALID_PARAMETER_VALUE),
+					errmsg("relation \"%s\" does not exist", name));
+		namestrcpy(&relinfo->attname, strVal(elem->arg));
+
+		relinfo->attnum = get_attnum(relinfo->hypertable_relid, NameStr(relinfo->attname));
+		if (relinfo->attnum == InvalidAttrNumber)
+			ereport(ERROR,
+					errcode(ERRCODE_INVALID_PARAMETER_VALUE),
+					errmsg("attribute \"%s\" does not exist", NameStr(relinfo->attname)));
+
+		relinfo_list = lappend(relinfo_list, relinfo);
+	}
+	return relinfo_list;
+}
+
+static void
+invalidations_write_record(HypertableInvalidationCacheEntry *entry, bool last_write,
+						   InvalidationsContext *args)
+{
+	LogicalDecodingContext *ctx = args->ctx;
+	InvalidationsPluginData *data = ctx->output_plugin_private;
+	TupleDesc tupdesc = RelationGetDescr(data->logrel);
+	Datum values[_Anum_continuous_aggs_hypertable_invalidation_log_max];
+	bool isnull[_Anum_continuous_aggs_hypertable_invalidation_log_max] = { false };
+
+	TS_DEBUG_LOG("write entry: hypertable_id=%u, lowest_modified_value=%" PRId64 ", "
+				 "greatest_modified_value=%" PRId64,
+				 entry->hypertable_relid,
+				 entry->lowest_modified_value,
+				 entry->greatest_modified_value);
+
+	/* Copy over the field to the invalidation slot */
+	values[AttrNumberGetAttrOffset(
+		Anum_continuous_aggs_hypertable_invalidation_log_hypertable_id)] =
+		ObjectIdGetDatum(entry->hypertable_relid);
+	values[AttrNumberGetAttrOffset(
+		Anum_continuous_aggs_hypertable_invalidation_log_lowest_modified_value)] =
+		Int64GetDatum(entry->lowest_modified_value);
+	values[AttrNumberGetAttrOffset(
+		Anum_continuous_aggs_hypertable_invalidation_log_greatest_modified_value)] =
+		Int64GetDatum(entry->greatest_modified_value);
+
+	MinimalTuple tuple = heap_form_minimal_tuple(tupdesc, values, isnull);
+	ExecStoreMinimalTuple(tuple, args->slot, true);
+
+	OutputPluginPrepareWrite(ctx, last_write);
+	/* Since we are not writing streamed data, we do not pass in an XID */
+	logicalrep_write_insert(ctx->out,
+							InvalidTransactionId,
+							data->logrel,
+							args->slot,
+							true, /* binary */
+							NULL);
+	OutputPluginWrite(ctx, last_write);
+}
+
+static void
+invalidations_startup(LogicalDecodingContext *ctx, OutputPluginOptions *opt, bool is_init)
+{
+	InvalidationsPluginData *data = palloc0(sizeof(InvalidationsPluginData));
+
+	data->context = AllocSetContextCreate(ctx->context,
+										  "Continuous aggregates invalidations memory context",
+										  ALLOCSET_DEFAULT_SIZES);
+	ctx->output_plugin_private = data;
+
+	opt->output_type = OUTPUT_PLUGIN_BINARY_OUTPUT;
+
+	if (!is_init)
+	{
+		List *relinfo_list = parse_output_parameters(ctx->output_plugin_options);
+
+		data->logrel =
+			table_openrv(makeRangeVar(CATALOG_SCHEMA_NAME,
+									  CONTINUOUS_AGGS_HYPERTABLE_INVALIDATION_LOG_TABLE_NAME,
+									  -1),
+						 AccessShareLock);
```

Comment:
Aren't we writing to this log table? So the lock should be RowExclusiveLock, no?
```suggestion
						 RowExclusiveLock);
```